src/services/ReportService.ts

```diff
@@ -0,0 +1,169 @@
+import { Service } from "typedi";
+import { EntityManager, Not } from "typeorm";
+import { InjectManager } from "typeorm-typedi-extensions";
+import { UserModel } from "../models/UserModel";
+import { PostModel } from "../models/PostModel";
+import { MessageModel } from "../models/MessageModel";
+import { ReportModel } from "../models/ReportModel";
+import { ReportPostRequest, ReportProfileRequest, ReportMessageRequest } from "../types/ApiRequests";
+import Repositories, { TransactionsManager } from "../repositories";
+import { report } from "process";
+import { UuidParam } from "../api/validators/GenericRequests";
+import { NotFoundError, ForbiddenError, UnauthorizedError } from "routing-controllers";
+
+@Service()
+export class ReportService {
+  private transactions: TransactionsManager;
+
+  constructor(@InjectManager() entityManager: EntityManager) {
+    this.transactions = new TransactionsManager(entityManager);
+  }
+
+  public async getAllReports(user: UserModel): Promise<ReportModel[]> {
+    return this.transactions.readOnly(async (transactionalEntityManager) => {
+      if (!user.admin) {
+        throw new Error("User does not have permission to get all reports");
+      }
+      return Repositories.report(transactionalEntityManager).getAllReports();
+    });
+  }
+
+  public async getAllPostReports(user: UserModel): Promise<ReportModel[]> {
+    return this.transactions.readOnly(async (transactionalEntityManager) => {
+      if (!user.admin) {
+        throw new Error("User does not have permission to get all post reports");
+      }
+      return Repositories.report(transactionalEntityManager).getAllPostReports();
+    });
+  }
+
+  public async getAllProfileReports(user: UserModel): Promise<ReportModel[]> {
+    return this.transactions.readOnly(async (transactionalEntityManager) => {
+      return Repositories.report(transactionalEntityManager).getAllProfileReports();
+    });
+  }
+
+  public async getAllMessageReports(user: UserModel): Promise<ReportModel[]> {
+    return this.transactions.readOnly(async (transactionalEntityManager) => {
+      return Repositories.report(transactionalEntityManager).getAllMessageReports();
+    });
```

Comment:
Out of curiosity, are these transaction wrappers necessary? I don't think we run any concurrent transactions so I'm unsure if we need these transaction isolation levels here, it's fine to keep them as they are though

Reply:
Pretty sure they are necessary. The whole codebase is written like this.